src/main/java/com/mraof/minestuck/network/EditmodeDragPacket.java

```diff
@@ -0,0 +1,281 @@
+						SoundType soundType = ((BlockItem) stack.getItem()).getBlock().defaultBlockState().getSoundType();
+						player.getLevel().playSound(player, pos, soundType.getPlaceSound(), SoundSource.BLOCKS, (soundType.getVolume() + 1.0F) / 2.0F, soundType.getPitch() * 0.8F);
+						
+						anyBlockPlaced = true;
+					}
+					else if (!conditionAndCostPair.getB().isEmpty())
+						missingCost.addGrist(conditionAndCostPair.getB());
+				}
+				
+				if(anyBlockPlaced)
+				{
+					//broadcasts edit sound to other players.
+					player.getLevel().playSound(player, positionEnd, MSSoundEvents.EVENT_EDIT_TOOL_REVISE.get(), SoundSource.AMBIENT, 1.0f, 1.0f);
+					player.swing(hand);
+				}
+				
+				if(!missingCost.isEmpty())
+					player.sendSystemMessage(missingCost.createMissingMessage(), true);
+				
+				ServerEditHandler.removeCursorEntity(player, !anyBlockPlaced);
+			}
+		}
+	}
+	
+	public record Destroy(boolean isDown, BlockPos positionStart, BlockPos positionEnd, Vec3 hitVector, Direction side) implements PlayToServerPacket
+	{
+		@Override
+		public void encode(FriendlyByteBuf buffer)
+		{
+			buffer.writeBoolean(isDown);
+			buffer.writeBlockPos(positionStart);
+			buffer.writeBlockPos(positionEnd);
+			buffer.writeDouble(hitVector.x);
+			buffer.writeDouble(hitVector.y);
+			buffer.writeDouble(hitVector.z);
+			buffer.writeEnum(side);
+		}
+		
+		public static Destroy decode(FriendlyByteBuf buffer)
+		{
+			boolean isDragging = buffer.readBoolean();
+			BlockPos positionStart = buffer.readBlockPos();
+			BlockPos positionEnd = buffer.readBlockPos();
+			Vec3 hitVector = new Vec3(buffer.readDouble(), buffer.readDouble(), buffer.readDouble());
+			Direction side = buffer.readEnum(Direction.class);
+			
+			return new Destroy(isDragging, positionStart, positionEnd, hitVector, side);
+		}
+		
+		@Override
+		public void execute(ServerPlayer player)
+		{
+			EditData data = ServerEditHandler.getData(player);
+			if(!player.getLevel().isClientSide() && data != null)
+			{
+				IEditTools cap = player.getCapability(MSCapabilities.EDIT_TOOLS_CAPABILITY).orElseThrow(() -> LOGGER.throwing(new IllegalStateException("EditTool Capability is missing on player " + player.getDisplayName().getString() + " on server-side (during packet execution)!")));
+				
+				cap.setEditPos1(positionStart);
+				cap.setEditPos2(positionEnd);
+				cap.setEditTrace(hitVector, side);
+				
+				GristSet missingCost = new GristSet();
+				boolean anyBlockDestroyed = false;
+				for(BlockPos pos : BlockPos.betweenClosed(positionStart, positionEnd))
+				{
+					Tuple<Boolean, GristSet> conditionAndCostPair = editModeDestroyCheck(data, player, pos);
+					if(conditionAndCostPair.getA())
+					{
+						player.gameMode.destroyAndAck(pos, 3, "creative destroy");
+						
+						//broadcasts block-break particles and sounds to other players.
+						player.level.levelEvent(2001, pos, Block.getId(player.getLevel().getBlockState(pos)));
```

Comment:
Are you getting the blockstate from the level after destroying it? Perhaps you should get the block state once at the start of the loop and then pass it around.

Reply:
That is a good point, I totally overlooked that.